src/hotspot/cpu/x86/x86.ad

```diff
@@ -8407,6 +8408,57 @@ instruct vmaskcast(vec dst) %{
   ins_pipe(empty);
 %}
 
+instruct vmaskcast_avx(vec dst, vec src) %{
+  predicate(Matcher::vector_length(n) == Matcher::vector_length(n->in(1)) &&
+            Matcher::vector_length_in_bytes(n) != Matcher::vector_length_in_bytes(n->in(1)) &&
+            (UseAVX >= 2 ||
+             (Matcher::vector_length_in_bytes(n) != 32 &&
+              Matcher::vector_length_in_bytes(n->in(1)) != 32)));
+  match(Set dst (VectorMaskCast src));
+  format %{ "vector_mask_cast $dst, $src" %}
+  ins_encode %{
+    int vlen = Matcher::vector_length(this);
+    BasicType src_bt = Matcher::vector_element_basic_type(this, $src);
+    BasicType dst_bt = Matcher::vector_element_basic_type(this);
+    __ vector_mask_cast($dst$$XMMRegister, $src$$XMMRegister, dst_bt, src_bt, vlen);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vmaskcast_avx1_32B_expand(vec dst, vec src, vec xtmp1, vec xtmp2) %{
+  predicate(UseAVX == 1 && Matcher::vector_length_in_bytes(n) == 32 &&
+            Matcher::vector_length(n) == Matcher::vector_length(n->in(1)) &&
+            Matcher::vector_length_in_bytes(n) > Matcher::vector_length_in_bytes(n->in(1)));
+  match(Set dst (VectorMaskCast src));
+  effect(TEMP xtmp1, TEMP xtmp2);
+  format %{ "vector_mask_cast $dst, $src\t! using $xtmp1, $xtmp2 as TEMP" %}
+  ins_encode %{
+    int vlen = Matcher::vector_length(this);
+    BasicType src_bt = Matcher::vector_element_basic_type(this, $src);
+    BasicType dst_bt = Matcher::vector_element_basic_type(this);
+    __ vector_mask_cast($dst$$XMMRegister, $src$$XMMRegister, $xtmp1$$XMMRegister,
+                                 $xtmp2$$XMMRegister, dst_bt, src_bt, vlen);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vmaskcast_avx1_32B_shrink(vec dst, vec src, vec xtmp) %{
+  predicate(UseAVX == 1 && Matcher::vector_length_in_bytes(n->in(1)) == 32 &&
```

Comment:
Limitation related to VectorLoadMask https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86.ad#L1852 will prevent exercising this pattern.